src/dbnode/storage/series/buffer.go

```diff
@@ -472,122 +448,360 @@ func (b *dbBuffer) FetchBlocksMetadata(
 	ctx context.Context,
 	start, end time.Time,
 	opts FetchBlocksMetadataOptions,
-) block.FetchBlockMetadataResults {
+) (block.FetchBlockMetadataResults, error) {
 	blockSize := b.opts.RetentionOptions().BlockSize()
 	res := b.opts.FetchBlockMetadataResultsPool().Get()
-	b.forEachBucketAsc(func(bucket *dbBufferBucket) {
-		if !bucket.canRead() {
-			return
+
+	for _, blockStart := range b.inOrderBlockStarts {
+		if !blockStart.Before(end) || !start.Before(blockStart.Add(blockSize)) {
+			continue
 		}
-		if !start.Before(bucket.start.Add(blockSize)) || !bucket.start.Before(end) {
-			return
+
+		bv, exists := b.bucketVersionsAt(blockStart)
+		if !exists {
+			// Invariant violated. This means the keys in the bucket map does
+			// not match the sorted keys cache, which should never happen.
+			log := b.opts.InstrumentOptions().Logger()
+			log.Errorf(errBucketMapCacheNotInSync)
```

Comment:
Here too, can you use the invariant violated method `EmitAndLogInvariantViolation(...)` in m3/x?
https://github.com/m3db/m3/blob/2228f1327a66bc9b0fcbc360b509099e382d8ba9/src/x/instrument/invariant.go#L71

That will guarantee a metric is emitted.